fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java

```diff
@@ -1217,20 +1219,50 @@ public static OptExpression cloneExpression(OptExpression logicalTree) {
      * @return : true if opt expression or its children have applied mv union rewrite, false otherwise.
      */
     public static boolean isAppliedMVUnionRewrite(OptExpression optExpression) {
+        return isOptHasAppliedRule(optExpression, Operator.OP_UNION_ALL_BIT);
+    }
+
+    public static boolean isOpAppliedRule(Operator op, int ruleMask) {
+        // TODO: support cte inline
+        int opRuleMask = op.getOpRuleMask();
+        return (opRuleMask & ruleMask) != 0;
+    }
+
+    public static boolean isOptHasAppliedRule(OptExpression optExpression, int ruleMask) {
         if (optExpression == null) {
             return false;
         }
-        if (optExpression.getOp().isOpRuleMaskSet(Operator.OP_UNION_ALL_BIT)) {
+        if (isOpAppliedRule(optExpression.getOp(), ruleMask)) {
             return true;
         }
         for (OptExpression child : optExpression.getInputs()) {
-            if (isAppliedMVUnionRewrite(child)) {
+            if (isOptHasAppliedRule(child, ruleMask)) {
                 return true;
             }
         }
         return false;
     }
 
+    /**
+     * Return mv's plan context. If mv's plan context is not in cache, optimize it.
+     * @param connectContext: connect context
+     * @param mv: input mv
+     * @param isInlineView: whether to inline mv's difined query.
+     */
+    public static MvPlanContext getMVPlanContext(ConnectContext connectContext,
+                                                 MaterializedView mv,
+                                                 boolean isInlineView) {
+        // step1: get from mv plan cache
+        List<MvPlanContext> mvPlanContexts = CachingMvPlanContextBuilder.getInstance()
+                .getPlanContext(mv, connectContext.getSessionVariable().isEnableMaterializedViewPlanCache());
+        if (mvPlanContexts != null && !mvPlanContexts.isEmpty() && mvPlanContexts.get(0).getLogicalPlan() != null) {
+            // TODO: distinguish normal mv plan and view rewrite plan
+            return mvPlanContexts.get(0);
```

Comment:
why?

Reply:
`getPlanContext` returns two plans, 1: inlined with view, 2: non-inlined with view.

Text based rewrite will only use the first one.